### PR TITLE
Remove references to torchtext.legacy from PyTorch Lightning

### LINKED ISSUE #10597: Deprecate support for torchtext.legacy dependencies
## Proposed refactor

Restrict checks & support for `torchtext.legacy` in Lightning

### Motivation

`torchtext` is retiring support for `torchtext.legacy` in upcoming versions: https://github.com/pytorch/text/issues/1422

Lightning uses the `Batch` from `torchtext.legacy` abstractions as part of this utility to move data to device:
https://github.com/PyTorchLightning/pytorch-lightning/blob/7f95fd04d77e3590c8073679e499eabb0c51f73c/pytorch_lightning/utilities/apply_func.py#L245-L255

### Pitch

Restrict the checks for `Batch` from torchtext based on the last torchtext version which supports them (torchtext <= 0.11).
Might be as simple as updating this line:
https://github.com/PyTorchLightning/pytorch-lightning/blob/7f95fd04d77e3590c8073679e499eabb0c51f73c/pytorch_lightning/utilities/apply_func.py#L29

to
```python
    if _compare_version("torchtext", operator.ge, "0.9.0") and _compare_version("torchtext", operator.le, "0.11"):
```
@parmeet @abhinavarora could you confirm?

Ultimately we could remove these checks as the torchtext team is not planning to support them. This would likely provide a performance boost on a critical path by removing a conditional check for torchtext when moving data to device.

### Additional context

cc @justusschock @awaelchli @akihironitta @tchaton @parmeet @abhinavarora 

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by abhinavarora**
## What does this PR do?

This PR resolves #10597 by deprecating references to `torchtext.legacy` from Lightning. As discussed in the issue, we add version checks for now to load the torchtext legacy components. In future, this code should be completely removed. 

Fixes #10597

### Does your PR introduce any breaking changes? If yes, please list them.
No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/imports.py
Comment (awaelchli) [inline pytorch_lightning/utilities/imports.py:97]: should the goal of this PR be that Lightning will warn users that have a version greater than 0.11? 
If we deprecate Lightning's support for the legacy torchtext Batch, we would want to print a message to the users, right? i.e., deprecate now in v1.6 and remove support in 1.8?

File: pytorch_lightning/utilities/apply_func.py
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:29]: this line should be replaced with the legacy check

File: pytorch_lightning/utilities/imports.py
Comment (carmocca) [reply pytorch_lightning/utilities/imports.py:97]: can we have instead

```suggestion
_TORCHTEXT_LEGACY = _TORCHTEXT_AVAILABLE and _compare_version("torchtext", operator.gt, "0.11.0")
```

Also, why is the version 0.11 and not 0.9? isn't 0.9 where it was deprecated?

File: tests/models/test_gpu.py
Comment (carmocca) [inline tests/models/test_gpu.py:312]: this should early return, instead of indenting everything

File: tests/utilities/test_apply_func_torchtext.py
Comment (carmocca) [inline tests/utilities/test_apply_func_torchtext.py:59]: You can use `@pytest.mark.skipif(_LEGACY_DEPRECATED)` instead

File: pytorch_lightning/utilities/imports.py
Comment (abhinavarora) [reply pytorch_lightning/utilities/imports.py:97]: @carmocca I agree with your suggestion. Also the version is 0.11 because we (torchtext) remove it in 0.11. 0.9 was when we migrated it tot `torchtext.legacy`.

@awaelchli Regarding your comment, I am not sure what is the best practice in Lightning to deprecate items. Either of these work fine for me. What is your recommendation. If you want to show a warning, what is the best place/way to do that ?

File: pytorch_lightning/utilities/apply_func.py
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:260]: I propose to emit a warning here, when user has a legacy Batch object, i.e., something like:

```py
rank_zero_warn("the torchtext.legacy.Batch object is deprecated and Lightning will remove support for it in v1.8. We recommend you to blablabla.")
```

File: pytorch_lightning/utilities/imports.py
Comment (awaelchli) [reply pytorch_lightning/utilities/imports.py:97]: I left a comment where I think the warning should be displayed https://github.com/PyTorchLightning/pytorch-lightning/pull/10724#discussion_r756469670

File: pytorch_lightning/utilities/apply_func.py
Comment (carmocca) [reply pytorch_lightning/utilities/apply_func.py:260]: Why would we want that?

Lightning is supposed to abstract this from the user, and select the correct class depending on the version.

File: tests/helpers/imports.py
Comment (carmocca) [inline tests/helpers/imports.py:3]: Dont' we want something like

if _SHOULD_USE_TORCHTEXT_LEGACY  # checks torchtext is <0.11
    from torchtext.legacy.data import Batch, Dataset, Example, Field, Iterator, LabelField
else:
    from torchtext.data import Batch, Dataset, Example, Field, Iterator, LabelField

File: pytorch_lightning/utilities/apply_func.py
Comment (awaelchli) [reply pytorch_lightning/utilities/apply_func.py:260]: My understanding is: There is no Batch class anymore in the latest torchtext. We want to remove the code entirely such that there is no torchtext Batch import anymore at all. In order to remove this support, we want to deprecate it such that users who user a newer version of Lightning and an older version of torchtext can make the transition.

@abhinavarora is my understanding about torchtext Batch correct?

Comment (abhinavarora) [reply pytorch_lightning/utilities/apply_func.py:260]: Hi @awaelchli, yes your understanding about torchtext Batch is correct.

File: tests/helpers/imports.py
Comment (abhinavarora) [reply tests/helpers/imports.py:3]: @carmocca , No we don't want this. Basically, this is the current status:
- In Torchtext <0.9, these classes were in `torchtext.data` package
- In Torchtext >0.9, these classes are in `torchtext.legacy.data` package
- After Torchtext 0.11, these classes will be completely removed. 

Hope that makes it clear?

Comment (abhinavarora) [reply tests/helpers/imports.py]: @carmocca , No we don't want this. Basically, this is the current status:

In Torchtext <0.9, these classes were in torchtext.data package
In Torchtext >0.9, these classes are in torchtext.legacy.data package
After Torchtext 0.11, these classes will be completely removed.

Hope that makes it clear?

Comment (carmocca) [reply tests/helpers/imports.py]: > After Torchtext 0.11, these classes will be completely removed.

From both `.data` and `.legacy.data`?

Comment (abhinavarora) [reply tests/helpers/imports.py]: @carmocca  Yes from both

File: pytorch_lightning/utilities/apply_func.py
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:32]: The

```python
else:
   Batch = type(None)
```

below can be removed right? we only access it if `_TORCHTEXT_LEGACY is True`

File: pytorch_lightning/utilities/apply_func.py
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:264]: Missing whitespace
```suggestion
                " We recommend you to migrate away from Batch by following the TorchText README: https://github.com/pytorch/text#bc-breaking-legacy"
```

File: pytorch_lightning/utilities/apply_func.py
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:261]: I believe we will be able to drop the dependency entirely, can you add a comment?

```suggestion
        if _TORCHTEXT_LEGACY and isinstance(data, Batch):
            # TODO: also remove the torchtext dependency with Lightning 1.8
```

File: pytorch_lightning/utilities/apply_func.py
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:262]: We have this for deprecations

```suggestion
            rank_zero_deprecation(
```

File: pytorch_lightning/utilities/apply_func.py
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:265]: Can you add a test for this message in https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/deprecated_api/test_remove_1-8.py?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/apply_func.py
Comment (abhinavarora) [reply pytorch_lightning/utilities/apply_func.py:32]: Yeah we could potentially remove this and add conditional imports in the test files. Alternatively, we can just define them as None Types.

File: pytorch_lightning/utilities/apply_func.py
Comment (abhinavarora) [reply pytorch_lightning/utilities/apply_func.py:261]: Done

File: pytorch_lightning/utilities/apply_func.py
Comment (abhinavarora) [reply pytorch_lightning/utilities/apply_func.py]: Done

**Comment by abhinavarora** [comment]
@carmocca @awaelchli As suggested by you folks, I changed to `rank_zero_deprecation` from `rank_zero_warn`. It seems like the unit tests fail when they encounter the depreciation warning. What is your guidance to fix that?

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
>  It seems like the unit tests fail when they encounter the depreciation warning. What is your guidance to fix that?

This was because some tests emitted the deprecation warning. In our test suite, we require each deprecation warning to be caught explicitly. I added these missing calls and will check back once test run completes here.

**Review by Borda — APPROVED** [review]

**Comment by abhinavarora** [comment]
Seems like all tests passed. @Borda @awaelchli @carmocca @ananthsub Could any of you with write access merge these changes?

**Comment by awaelchli** [comment]
Yep, I can confirm everything is fine with this PR. We are currently investigating issues with our CI hanging tests which impacts this PR as well but is unrelated. Apologies and thanks for the patience. I will come back to the PR and rerun tests

**Comment by Nic-Ma** [comment]
Hi @awaelchli ,

Thanks for the quick fix, may I know which version will contain this fix?
We are blocked by this issue in MONAI:
https://github.com/Project-MONAI/tutorials/issues/530

Thanks in advance.

**Comment by Nic-Ma** [comment]
Hi @abhinavarora ,

Currently, I fixed the PyTorch version to 1.10.2 as a workaround in the tutorial and didn't see the issue anymore.
Will update it when you guys released a new version.

Thanks.

**Pull request merged**
